File: src-tauri/src/input_backends.rs
//! GameInputBackend — backends d'application du remapping clavier par jeu.
//!
//! Couche purement décisionnelle et en lecture seule (Phase 6). L'application
//! réelle (bindings du jeu, HKL, interception scancodes limitée à la fenêtre)
//! est fournie par des adaptateurs séparés, à tester sur un vrai jeu.
//! Pour Neverness to Everness (Anti-Cheat Expert) : aucun hook, injection,
//! driver, écriture processus ni modification des fichiers du jeu.
//!
//! Ce module doit rester synchronisé avec `src/lib/inputBackends.ts`.

use serde::{Deserialize, Serialize};

/// Identifiants des backends, dans l'ordre de préférence :
/// 1. bindings natifs du jeu ; 2. layout reconnu par le jeu ;
/// 3. remapping ZAILON limité à la fenêtre ; 4. Steam Input ;
/// 5. aucune méthode.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum InputBackendId {
    NativeBinding,
    WindowLayout,
    ScopedRemap,
    SteamInput,
    Unsupported,
}

impl InputBackendId {
    pub const PRIORITY: [InputBackendId; 5] = [
        InputBackendId::NativeBinding,
        InputBackendId::WindowLayout,
        InputBackendId::ScopedRemap,
        InputBackendId::SteamInput,
        InputBackendId::Unsupported,
    ];

    pub fn label(self) -> &'static str {
        match self {
            Self::NativeBinding => "Bindings natifs du jeu",
            Self::WindowLayout => "Layout reconnu par le jeu",
            Self::ScopedRemap => "Remapping ZAILON limité à la fenêtre",
            Self::SteamInput => "Steam Input",
            Self::Unsupported => "Aucune méthode",
        }
    }

    pub fn short_label(self) -> &'static str {
        match self {
            Self::NativeBinding => "Bindings natifs",
            Self::WindowLayout => "Layout Windows",
            Self::ScopedRemap => "Remap fenêtre",
            Self::SteamInput => "Steam Input",
            Self::Unsupported => "Aucune",
        }
    }
}

/// Portée d'application d'un backend.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum BackendScope {
    GameConfig,
    WindowsHkl,
    WindowOnly,
    Controller,
    None,
}

/// Métadonnées de sécurité et de portée d'un backend.
#[derive(Debug, Clone, Copy)]
pub struct InputBackendMeta {
    pub id: InputBackendId,
    pub scope: BackendScope,
    pub priority: u8,
    pub safe_with_anti_cheat: bool,
    pub requires_injection: bool,
    pub requires_driver: bool,
    pub requires_process_write: bool,
    pub touches_desktop: bool,
    pub windows_language_change: bool,
}

pub fn backend_meta(id: InputBackendId) -> InputBackendMeta {
    match id {
        InputBackendId::NativeBinding => InputBackendMeta {
            id,
            scope: BackendScope::GameConfig,
            priority: 1,
            safe_with_anti_cheat: false,
            requires_injection: false,
            requires_driver: false,
            requires_process_write: false,
            touches_desktop: false,
            windows_language_change: false,
        },
        InputBackendId::WindowLayout => InputBackendMeta {
            id,
            scope: BackendScope::WindowsHkl,
            priority: 2,
            safe_with_anti_cheat: true,
            requires_injection: false,
            requires_driver: false,
            requires_process_write: false,
            touches_desktop: true,
            windows_language_change: true,
        },
        InputBackendId::ScopedRemap => InputBackendMeta {
            id,
            scope: BackendScope::WindowOnly,
            priority: 3,
            safe_with_anti_cheat: true,
            requires_injection: false,
            requires_driver: false,
            requires_process_write: false,
            touches_desktop: false,
            windows_language_change: false,
        },
        InputBackendId::SteamInput => InputBackendMeta {
            id,
            scope: BackendScope::Controller,
            priority: 4,
            safe_with_anti_cheat: true,
            requires_injection: false,
            requires_driver: false,
            requires_process_write: false,
            touches_desktop: false,
            windows_language_change: false,
        },
        InputBackendId::Unsupported => InputBackendMeta {
            id,
            scope: BackendScope::None,
            priority: 5,
            safe_with_anti_cheat: true,
            requires_injection: false,
            requires_driver: false,
            requires_process_write: false,
            touches_desktop: false,
            windows_language_change: false,
        },
    }
}

pub fn is_nte(game_name: &str) -> bool {
    game_name.to_lowercase().contains("neverness")
}

/// Anti-Cheat détecté (lecture seule, par nom de jeu connu).
pub fn detect_anti_cheat(game_name: &str) -> Option<&'static str> {
    if is_nte(game_name) {
        Some("Anti-Cheat Expert")
    } else {
        None
    }
}

/// Disponibilité d'un backend pour un jeu donné, avec la raison.
#[derive(Debug, Clone)]
pub struct BackendAvailability {
    pub backend: InputBackendId,
    pub available: bool,
    pub reason: String,
}

/// Plan d'application : backend choisi + chaîne de repli + contraintes.
#[derive(Debug, Clone)]
pub struct InputBackendPlan {
    pub chosen: InputBackendId,
    pub chain: Vec<BackendAvailability>,
    pub anti_cheat: Option<&'static str>,
    pub constraints: Vec<String>,
}

/// Planifie la méthode d'application pour un jeu. Ne modifie rien.
pub fn plan_input_backend(game_name: &str, has_mapping: bool) -> InputBackendPlan {
    let anti_cheat = detect_anti_cheat(game_name);
    let nte = is_nte(game_name);
    let mut constraints = Vec::new();

    if let Some(name) = anti_cheat {
        constraints.push(format!(
            "{name} détecté : aucun hook, injection, driver ni manipulation du processus."
        ));
    }
    if nte {
        constraints.push(
            "Neverness to Everness : les retours joueurs montrent qu'un simple changement de \
             layout logique ne suffit pas (touches physiques/scancodes)."
                .to_string(),
        );
        constraints.push(
            "L'adaptateur NTE doit tester la méthode réellement fonctionnelle via la sonde de \
             compatibilité avant application."
                .to_string(),
        );
    }
    if !has_mapping {
        constraints
            .push("Aucune traduction active : la disposition s'applique telle quelle.".to_string());
    }

    let chain: Vec<BackendAvailability> = InputBackendId::PRIORITY
        .iter()
        .copied()
        .map(|id| {
            let mut available = true;
            let mut reason = String::new();

            match id {
                InputBackendId::NativeBinding => {
                    if nte {
                        available = false;
                        reason = "Interdit pour NTE : la modification des fichiers du jeu est \
                                  exclue avec Anti-Cheat Expert."
                            .to_string();
                    }
                }
                InputBackendId::WindowLayout => {
                    if nte {
                        available = false;
                        reason = "Inefficace constaté pour NTE (touches physiques) — la sonde la \
                                  teste néanmoins avant de conclure."
                            .to_string();
                    }
                }
                InputBackendId::ScopedRemap => {
                    // Toujours disponible : externe au jeu, sûr avec Anti-Cheat.
                }
                InputBackendId::SteamInput => {
                    if nte {
                        available = false;
                        reason = "Non pertinent pour NTE en clavier/souris ; réservé aux manettes \
                                  et uniquement si réellement compatible."
                            .to_string();
                    }
                }
                InputBackendId::Unsupported => {}
            }
            if !has_mapping && id != InputBackendId::Unsupported {
                available = false;
                reason = "Aucune traduction active à appliquer.".to_string();
            }
            if reason.is_empty() {
                reason = match id {
                    InputBackendId::NativeBinding => "Écrit la configuration d'entrée du jeu. À éviter sur les jeux protégés par Anti-Cheat sans confirmation.",
                    InputBackendId::WindowLayout => "Change temporairement le layout logique Windows. Inefficace pour les jeux qui lisent les touches physiques.",
                    InputBackendId::ScopedRemap => "Traduit les touches uniquement lorsque la fenêtre du jeu est au premier plan. Aucun hook, aucun fichier modifié, aucun driver.",
                    InputBackendId::SteamInput => "Pensé pour les manettes, pas pour clavier/souris. Ne jamais modifier une configuration Steam Input existante sans confirmation.",
                    InputBackendId::Unsupported => "Aucune application automatique : la disposition reste un choix manuel dans le jeu.",
                }
                .to_string();
            }
            BackendAvailability { backend: id, available, reason }
        })
        .collect();

    let chosen = chain
        .iter()
        .find(|item| item.available)
        .map(|item| item.backend)
        .unwrap_or(InputBackendId::Unsupported);

    InputBackendPlan {
        chosen,
        chain,
        anti_cheat,
        constraints,
    }
}

/// Sonde de compatibilité NTE (lecture seule).
///
/// Les champs de détection (processus, fenêtre au premier plan, layout Windows)
/// seront branchés sur `windows-sys` dans un second temps ; tant qu'ils ne le
/// sont pas, la sonde ne fait que refléter la configuration et n'applique rien.
#[derive(Debug, Clone)]
pub struct NteCompatibilityProbe {
    pub process_detected: bool,
    pub window_foreground: bool,
    pub windows_layout: String,
    pub anti_cheat: bool,
}

impl NteCompatibilityProbe {
    /// Décision de la sonde : ne jamais choisir un backend interdit par l'Anti-Cheat.
    pub fn plan(&self, game_name: &str, has_mapping: bool) -> InputBackendPlan {
        let mut plan = plan_input_backend(game_name, has_mapping);
        if self.anti_cheat {
            plan.constraints.push(
                "Anti-Cheat actif : la sonde reste en lecture seule, aucun test d'application."
                    .to_string(),
            );
        }
        plan
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn priority_order_is_stable() {
        assert_eq!(
            InputBackendId::PRIORITY,
            [
                InputBackendId::NativeBinding,
                InputBackendId::WindowLayout,
                InputBackendId::ScopedRemap,
                InputBackendId::SteamInput,
                InputBackendId::Unsupported,
            ]
        );
    }

    #[test]
    fn generic_game_prefers_native_bindings() {
        let plan = plan_input_backend("Cyberpunk 2077", true);
        assert_eq!(plan.chosen, InputBackendId::NativeBinding);
        assert!(plan.anti_cheat.is_none());
        assert!(plan.chain.iter().all(|item| item.available));
    }

    #[test]
    fn nte_with_mapping_chooses_scoped_remap() {
        let plan = plan_input_backend("Neverness to Everness", true);
        assert_eq!(plan.anti_cheat, Some("Anti-Cheat Expert"));
        assert_eq!(plan.chosen, InputBackendId::ScopedRemap);
        let native = plan
            .chain
            .iter()
            .find(|item| item.backend == InputBackendId::NativeBinding)
            .unwrap();
        assert!(!native.available);
        let window = plan
            .chain
            .iter()
            .find(|item| item.backend == InputBackendId::WindowLayout)
            .unwrap();
        assert!(!window.available);
        let steam = plan
            .chain
            .iter()
            .find(|item| item.backend == InputBackendId::SteamInput)
            .unwrap();
        assert!(!steam.available);
    }

    #[test]
    fn no_mapping_falls_back_to_unsupported() {
        let plan = plan_input_backend("Neverness to Everness", false);
        assert_eq!(plan.chosen, InputBackendId::Unsupported);
    }

    #[test]
    fn ace_probe_stays_read_only() {
        let probe = NteCompatibilityProbe {
            process_detected: true,
            window_foreground: true,
            windows_layout: "AZERTY".to_string(),
            anti_cheat: true,
        };
        let plan = probe.plan("Neverness to Everness", true);
        assert_eq!(plan.chosen, InputBackendId::ScopedRemap);
        assert!(plan
            .constraints
            .iter()
            .any(|item| item.contains("lecture seule")));
    }
}

File: src/lib/inputBackends.ts
import type { Game, GameKeyboardLayout } from '../types'

/**
 * GameInputBackend — architecture des backends d'entrée (Phase 6).
 *
 * Un backend est une méthode d'application de la disposition clavier par jeu.
 * L'ordre de préférence (spec « UI compacte + remapping NTE », §9) :
 *   1. bindings natifs du jeu ;
 *   2. changement de layout reconnu par le jeu ;
 *   3. remapping temporaire ZAILON limité à la fenêtre du jeu ;
 *   4. Steam Input uniquement lorsque réellement compatible ;
 *   5. aucune méthode.
 *
 * Cette couche TypeScript est purement décisionnelle et en lecture seule :
 * l'application réelle (bindings, HKL, interception scancodes) est fournie par
 * les backends Rust de la Phase 6, à tester sur un vrai jeu.
 */

export type InputBackendId =
  | 'native-binding'
  | 'window-layout'
  | 'scoped-remap'
  | 'steam-input'
  | 'unsupported'

export type BackendScope = 'game-config' | 'windows-hkl' | 'window-only' | 'controller' | 'none'

export interface InputBackend {
  id: InputBackendId
  label: string
  shortLabel: string
  scope: BackendScope
  priority: number
  safeWithAntiCheat: boolean
  requiresInjection: boolean
  requiresDriver: boolean
  requiresProcessWrite: boolean
  touchesDesktop: boolean
  windowsLanguageChange: boolean
  note: string
}

export const INPUT_BACKENDS: Record<InputBackendId, InputBackend> = {
  'native-binding': {
    id: 'native-binding',
    label: 'Bindings natifs du jeu',
    shortLabel: 'Bindings natifs',
    scope: 'game-config',
    priority: 1,
    safeWithAntiCheat: false,
    requiresInjection: false,
    requiresDriver: false,
    requiresProcessWrite: false,
    touchesDesktop: false,
    windowsLanguageChange: false,
    note: 'Écrit la configuration d’entrée du jeu. À éviter sur les jeux protégés par Anti-Cheat sans confirmation.',
  },
  'window-layout': {
    id: 'window-layout',
    label: 'Layout reconnu par le jeu',
    shortLabel: 'Layout Windows',
    scope: 'windows-hkl',
    priority: 2,
    safeWithAntiCheat: true,
    requiresInjection: false,
    requiresDriver: false,
    requiresProcessWrite: false,
    touchesDesktop: true,
    windowsLanguageChange: true,
    note: 'Change temporairement le layout logique Windows. Inefficace pour les jeux qui lisent les touches physiques/scancodes (ex. NTE).',
  },
  'scoped-remap': {
    id: 'scoped-remap',
    label: 'Remapping ZAILON limité à la fenêtre',
    shortLabel: 'Remap fenêtre',
    scope: 'window-only',
    priority: 3,
    safeWithAntiCheat: true,
    requiresInjection: false,
    requiresDriver: false,
    requiresProcessWrite: false,
    touchesDesktop: false,
    windowsLanguageChange: false,
    note: 'Traduit les touches uniquement lorsque la fenêtre du jeu est au premier plan. Aucun hook dans le jeu, aucun fichier modifié, aucun driver.',
  },
  'steam-input': {
    id: 'steam-input',
    label: 'Steam Input',
    shortLabel: 'Steam Input',
    scope: 'controller',
    priority: 4,
    safeWithAntiCheat: true,
    requiresInjection: false,
    requiresDriver: false,
    requiresProcessWrite: false,
    touchesDesktop: false,
    windowsLanguageChange: false,
    note: 'Pensé pour les manettes, pas pour clavier/souris. Ne jamais modifier une configuration Steam Input existante sans confirmation.',
  },
  unsupported: {
    id: 'unsupported',
    label: 'Aucune méthode',
    shortLabel: 'Aucune',
    scope: 'none',
    priority: 5,
    safeWithAntiCheat: true,
    requiresInjection: false,
    requiresDriver: false,
    requiresProcessWrite: false,
    touchesDesktop: false,
    windowsLanguageChange: false,
    note: 'Aucune application automatique : la disposition reste un choix manuel dans le jeu.',
  },
}

export const BACKEND_PRIORITY: InputBackendId[] = ['native-binding', 'window-layout', 'scoped-remap', 'steam-input', 'unsupported']

export const isNte = (gameName: string) => gameName.toLocaleLowerCase().includes('neverness')

export function detectAntiCheat(gameName: string): { detected: boolean; name?: string } {
  if (isNte(gameName)) return { detected: true, name: 'Anti-Cheat Expert' }
  return { detected: false }
}

export interface InputBackendAvailability {
  backend: InputBackend
  available: boolean
  reason: string
}

export interface InputBackendPlan {
  chosen: InputBackendId
  chain: InputBackendAvailability[]
  antiCheat: { detected: boolean; name?: string }
  constraints: string[]
}

/**
 * Planifie la méthode d'application pour un jeu, en lecture seule.
 * Ne modifie rien : pure décision à partir du nom du jeu et de la configuration.
 */
export function planInputBackend(game: Game, layout: GameKeyboardLayout): InputBackendPlan {
  const antiCheat = detectAntiCheat(game.name)
  const constraints: string[] = []
  const nte = isNte(game.name)
  const hasMapping = (game.keyboardProfiles || []).some(profile => profile.enabled && profile.mapping.length > 0)

  if (antiCheat.detected) {
    constraints.push(`${antiCheat.name} détecté : aucun hook, injection, driver ni manipulation du processus.`)
  }
  if (nte) {
    constraints.push('Neverness to Everness : les retours joueurs montrent qu’un simple changement de layout logique ne suffit pas (touches physiques).')
    constraints.push('L’adaptateur NTE doit tester la méthode réellement fonctionnelle via la sonde de compatibilité (Phase 6 Rust) avant application.')
  }
  if (!hasMapping) {
    constraints.push('Aucune traduction active : la disposition s’applique telle quelle, sans remapping.')
  }

  const chain: InputBackendAvailability[] = BACKEND_PRIORITY.map(id => {
    const backend = INPUT_BACKENDS[id]
    let available = true
    let reason = backend.note

    if (id === 'native-binding') {
      if (nte) {
        available = false
        reason = 'Interdit pour NTE : la modification des fichiers du jeu est exclue avec Anti-Cheat Expert.'
      }
    } else if (id === 'window-layout') {
      if (nte) {
        available = false
        reason = 'Inefficace constaté pour NTE (touches physiques/scancodes) — la sonde la teste néanmoins avant de conclure.'
      }
    } else if (id === 'scoped-remap') {
      // Toujours disponible : externe au jeu, sûr avec Anti-Cheat.
    } else if (id === 'steam-input') {
      if (nte) {
        available = false
        reason = 'Non pertinent pour NTE en clavier/souris ; réservé aux manettes et uniquement si réellement compatible.'
      }
    }
    if (!hasMapping && id !== 'unsupported') {
      available = false
      reason = 'Aucune traduction active à appliquer.'
    }
    return { backend, available, reason }
  })

  const chosen: InputBackendId = chain.find(item => item.available)?.backend.id || 'unsupported'

  return { chosen, chain, antiCheat, constraints }
}

export interface InputDiagnosticRow {
  label: string
  value: string
  tone?: 'ok' | 'warn' | 'neutral'
}

/**
 * État visible (spec §15/§20) — diagnostic d'entrée par jeu.
 * Les détections de processus/fenêtre sont fournies par la Phase 6 Rust ;
 * tant qu'elles n'existent pas, la colonne est honnêtement « non surveillé ».
 */
export function inputDiagnosticRows(game: Game, layout: GameKeyboardLayout): InputDiagnosticRow[] {
  const plan = planInputBackend(game, layout)
  const nte = isNte(game.name)
  return [
    { label: 'Jeu', value: game.name },
    { label: 'Disposition souhaitée', value: layout.toUpperCase() },
    { label: 'Méthode', value: INPUT_BACKENDS[plan.chosen].label },
    { label: 'État', value: plan.chosen === 'scoped-remap' ? 'Actif uniquement dans le jeu' : plan.chosen === 'unsupported' ? 'Aucune application automatique' : 'Selon backend', tone: plan.chosen === 'unsupported' ? 'warn' : 'ok' },
    { label: 'ACE détecté', value: plan.antiCheat.detected ? (plan.antiCheat.name || 'oui') : 'non', tone: plan.antiCheat.detected ? 'warn' : 'ok' },
    { label: 'Injection', value: 'non' },
    { label: 'Driver', value: 'non' },
    { label: 'Langue Windows ajoutée', value: 'non' },
    { label: 'Processus / Fenêtre', value: nte ? 'Sonde Phase 6 Rust (non surveillé ici)' : '—', tone: 'neutral' },
  ]
}
